Minimal boilerplate for micro services - Initial set of features

jifeon opened the pull request:
Roadmap:

- [x] configured eslint
- [x] precommit hook checking all modified js files with eslint and fixing code style with prettier
- [x] simple express application with error handling and example of API endpoint
- [x] Docker build for production
- [x] docker-compose for development
- [x] hot reload (implemented, need to check)
- [x] CircleCI config
- [x] Logger with support to pipe logs into elasticsearch
- [x] Docs

jifeon committed:
- configured eslint
- precommit hook checking all modified js files with eslint and fixing code style with prettier
- simple express application with example of API endpoint
- Docker build for production
- docker-compose for development
- hot reload
- CircleCI config
- Logger with support to pipe logs into elasticsearch

jifeon commented on src/app.js:
```diff
@@ -0,0 +1,51 @@
+  })
+);
+
+app.use("/", require("./api/hello"));
```
Thinking about moving it to `endpoints.js` since it's a part which should be edited by users of the boilerplate

apsavin commented on package.json:
```diff
@@ -0,0 +1,26 @@
+    "pino": "^4.17.3",
+    "pino-elasticsearch": "^2.2.1",
+    "pino-http": "^3.2.2",
+    "superagent": "^3.8.3"
```
Could you please provide some examples of microservices? Does a typical microservice send http requests?

jifeon replied:
Yes, for us it usually microservices which communicate by HTTP. First free real tasks are integrations with 3rd party API. The services communicate with these APIs periodically, format answers special way and send to our API.

apsavin replied:
Thanks!

apsavin commented on package.json:
```diff
@@ -0,0 +1,26 @@
+    "@plyo/dev-tools": "1.0.0-alpha.1"
+  },
+  "scripts": {
+    "lint": "node ./node_modules/@plyo/dev-tools/run/lint.js",
```
Can't we use `@plyo/dev-tools/run/lint.js` like `eslint` here, without `node ./node_modules/`?

BMalaichik replied:
we can modify package.json entry:

https://docs.npmjs.com/files/package.json#bin

apsavin commented on src/middlewares/errorHandler.js:
```diff
@@ -0,0 +1,85 @@
+  }
+}
+
+module.exports = errorHandler;
```
Looks like a candidate for `@plyo/errorHandlerMiddleware` ;)

BMalaichik replied:
even `@plyo/http` standalone package

jifeon replied:
yep, thought about it as well, we have  set of errors as well, which is not included yet. Maybe use `plyo-tools` for now since it already exists and contains errors?

wanchopen commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Minimal boilerplate for micro services](https://trello.com/c/t0DkB0MW/1687-minimal-boilerplate-for-micro-services)

BMalaichik commented on src/server.js:
```diff
@@ -0,0 +1,67 @@
+    if (message === "reload") {
+      reload = reload.then(() => shutDown()).then(() => {
+        Object.keys(require.cache).forEach(key => {
+          if (key.indexOf("node_modules") === -1) delete require.cache[key];
```
why do you manually invalidate cache??? I am not sure it's a good idea..
for such restart purposes simple pkg [nodemon](https://www.npmjs.com/package/nodemon).

I think better to delegate restart init to this tool

jifeon replied:
this approach we use for very long time for `plyo.db`, it was part of node-starter-kit for some time I believe. It worked perfect, but I need to test how it works in boilerplate. Looks like I forgot about some dependencies. I'll think about alternatives. Meanwhile @frenzzy what's you opinion on this question?

jifeon replied:
Actually I forgot to move [one more file](https://github.com/plyo/plyo.db/blob/master/scripts/builder/run.js), and I start thinking that we should switch to nodemon, since we going to avoid build stage in micro services and can keep them really simple.

jifeon replied:
@frenzzy reminded me how nodemon works, and it's much slower. It's not critical for small projects, but plyo.db initialization took a lot of time and we want to keep this process similar for all the projects.

BMalaichik replied:
@jifeon slow? hm, interesting. didn't face such problems with it before.
just curios about circumstances where it has been slow

BMalaichik reviewed (commented):
Looks good!

JFYI, did you take a look at https://www.fastify.io/ ?

looks quite promising: faster, easy-to-go & familiar API.

jifeon commented:
> JFYI, did you take a look at https://www.fastify.io/ ?

For now express itself is not a bottleneck in our projects, so I would like to stick with it. But thanks for the link, I'll keep in mind.

VlVl reviewed (commented):
LGTM

jifeon committed:
hot reload replaced with watchmen

jifeon commented on entrypoint.sh:
```diff
@@ -0,0 +1,14 @@
+cat ${pipe} | ./node_modules/.bin/pino &
+
+# tee forks logs into `$pipe` and elasticsearch
+node build/server.js | tee ${pipe} | ./node_modules/.bin/pino-elasticsearch --host ${ELASTIC_HOST}
```
oops there is no `build` folder, need to check production-like run

jifeon commented on package.json:
```diff
@@ -0,0 +1,27 @@
+    "lint": "lint",
+    "lint-staged": "lint --staged",
+    "precommit": "lint --staged",
+    "start": "nodemon src/server.js"
```
I've checked nodemon for plyo.db and found out that there is no difference in performance, both solution works fast "enough". Looks like it's because approach for plyo.db should be refactored to not rerun the whole server. But for simplicity of microservices, I've added `nodemon` into `@plyo/dev-tools`

frenzzy replied:
IMO, if you are using some module explicitly, you should add it to dependencies list.
It is a very minor. Another solution is something like this:
```js
"start": "@plyo/dev-tools start src/server.js"
```

jifeon replied:
I like proposed solution, but how would you implement it? Is it a magic way to run `start` script from `@plyo/dev-tools`?

frenzzy replied:
Example: https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/bin/react-scripts.js

```js
// @plyo/dev-tools/tasks/start.js
const nodemon = require('nodemon')
nodemon({ script: process.argv[2] })
```

jifeon commented on package.json:
```diff
@@ -0,0 +1,27 @@
+  "scripts": {
+    "lint": "lint",
+    "lint-staged": "lint --staged",
+    "precommit": "lint --staged",
```
@BMalaichik @apsavin fixed with `bin`, thank you, looks cleaner now

frenzzy replied:
What is `lint`? Should not we be more explicit, Why not `@plyo/dev-tools lint` or similar?

jifeon commented on src/config/config.js:
```diff
@@ -0,0 +1,18 @@
+module.exports = {
+  app: {
+    port: process.env.PORT || 3000,
+    keepAliveConnection: process.env.KEEP_ALIVE_CONNECTION || false,
```
it's outdated now

jifeon commented on src/config/config.js:
```diff
@@ -0,0 +1,18 @@
+  },
+
+  logger: {
+    name: process.env.ENV_NAME || "local",
```
need to add this env var into docker-compose

jifeon commented on src/server.js:
```diff
@@ -0,0 +1,27 @@
+  logger.info(`The server is running at http://localhost:${port}/`);
+});
+
+module.exports = server;
```
I think this is as simple as possible now

frenzzy commented on Dockerfile:
```diff
@@ -0,0 +1,10 @@
+WORKDIR /usr/src/app/
+
+COPY package.json yarn.lock /usr/src/app/
+RUN yarn install --production && rm -rf /usr/local/share/.cache
```
What the `rm -rf /usr/local/share/.cache` actually do? Why it happens after install?

jifeon replied:
It helps to reduce the size of docker image dramatically by removing yarn cache. You can read more details in this card https://trello.com/c/m18M3j1k/1443-fix-dev-server. I'll put inline comments about it

BMalaichik reviewed (approved):
looks cool!

jifeon committed:
nodemon watch fixed, bnluebird added

jifeon committed:
prettier config added

jifeon committed:
- CircleCI build added
- docker-compose.override added to .gitignore
- dockerfile and entrypoint typos fixed
- dev-tools updated

jifeon committed:
docs for docker-compose

jifeon commented:
Ok, I think it's ready for now. Other changes will be delivered as separate PRs
